CARDS-59: Add support for running frontend tests using Selenium

acrowthe opened the pull request:
Add selenium test module. This module can be run using `mvn clean install -Pdocker -Dselenium`. It runs tests using Docker to launch CARDS and Google Chrome to run the tests. To support this, `-Dselenium` will add docker to the build in addition to cleaning out any running or old CARDS docker builds.

To run these tests, make sure that you have Google Chrome installed in the default install location for your operating system. The expected location can be found here:
https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver/01fde32d0ed245141e24151f83b7c2db31d596a4#requirements

If you are using WSL, make sure that Google Chrome is installed through WSL, not just through Windows:
https://scottspence.com/posts/use-chrome-in-ubuntu-wsl

The current test is to login and verify that the user's avatar on the top right of the dashboard is correct. To ensure that this test ran, you can check `/selenium/target/failsafe-reports`. There should be 1 run test and no failures.

If you want to check behaviour if there is a failed test, modify `LoginSelenium.java` and change `assertEquals("A"` to some other string.

Framework has been added for supporting other browsers though there are issues with browsers other than chrome:
- Firefox does is not waiting for user login to complete before checking if the user has logged in
- Edge is launching Chrome and then complaining that Chrome is not Edge
- Safari framework is in place but untested

When no browser is provided, testing will default to Chrome

To launch with other browsers, make sure they are installed in the default install location for your operating system then run
`mvn clean install -Pdocker -Dselenium -DseleniumDriver=<browser>`
where `<browser>` is one of `edge`, `chrome`, `firefox` or `safari`

acrowthe committed:
CARDS-59: Add support for running frontend tests using Selenium
Add selenium test module. This module can be run using `mvn clean
install -Dselenium`. It runs tests using Docker to launch CARDS and
Google Chrome to run the tests. To support this, `-Dselenium` will
add docker to the build in addition to cleaning out any running or
old CARDS docker builds.

To run these tests, make sure that you have Google Chrome installed
in the default install location for your operating system. The
expected location can be found here:
https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver/01fde32d0ed245141e24151f83b7c2db31d596a4#requirements

If you are using WSL, make sure that Google Chrome is installed
through WSL, not just through Windows:
https://scottspence.com/posts/use-chrome-in-ubuntu-wsl

The current test is to login and verify that the user's avatar on
the top right of the dashboard is correct. To ensure that this test
ran, you can check `/selenium/target/failsafe-reports`. There should
be 1 run test and no failures.

If you want to check behaviour if there is a failed test, modify
`LoginSelenium.java` and change `assertEquals("A"` to some other
string.

IntegralProgrammer commented on pom.xml:
```diff
@@ -389,7 +399,6 @@
       <plugin>
         <artifactId>maven-compiler-plugin</artifactId>
         <configuration>
-          <compilerArgument>-Xlint:all</compilerArgument>
```
Why is this removed?

IntegralProgrammer commented on selenium/launch_cards.sh:
```diff
@@ -0,0 +1,48 @@
+# specific language governing permissions and limitations
+# under the License.
+
+CARDS_PID=$(pgrep "start_cards.sh")
```
This `CARDS_PID` value doesn't appear to be used anywhere.

IntegralProgrammer commented on selenium/launch_cards.sh:
```diff
@@ -0,0 +1,48 @@
+CARDS_PID=$(pgrep "start_cards.sh")
+
+function check_cards_running() {
+  docker ps | grep compose-cluster-cardsinitial-1 > /dev/null
```
I'm not sure if this perhaps varies across Docker / Docker Compose versions but on the version that I am using (`apt install docker-compose` on Debian 11), the running Docker container is listed as `compose-cluster_cardsinitial_1` as opposed to `compose-cluster-cardsinitial-1`. This resulted in a failure when I tried to run the tests with `mvn clean install -Dselenium`. I edited this `launch_cards.sh` file and changed `compose-cluster-cardsinitial-1` to `compose-cluster_cardsinitial_1` and the test passed. Perhaps a more robust solution would be to do something like `docker inspect $(docker-compose ps -q cardsinitial) --format "{{.State.Status}}" | grep "running"`.

IntegralProgrammer commented on selenium/launch_cards.sh:
```diff
@@ -0,0 +1,48 @@
+}
+
+function handle_cards_java_fail() {
+  echo -e "Cards has failed to start, cannot run selenium tests"
```
```suggestion
  echo -e "CARDS has failed to start, cannot run Selenium tests"
```

IntegralProgrammer commented on selenium/launch_cards.sh:
```diff
@@ -0,0 +1,48 @@
+  exit -1
+}
+
+echo "Cleaning up old CARDS docker images"
```
```suggestion
echo "Cleaning up old CARDS Docker Compose deployments"
```

IntegralProgrammer commented on selenium/launch_cards.sh:
```diff
@@ -0,0 +1,48 @@
+
+echo "Cleaning up old CARDS docker images"
+docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh
+echo "Generating docker configuration"
```
```suggestion
echo "Generating CARDS Docker Compose environment"
```

IntegralProgrammer commented on selenium/launch_cards.sh:
```diff
@@ -0,0 +1,48 @@
+docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh
+echo "Generating docker configuration"
+python3 generate_compose_yaml.py --oak_filesystem --dev_docker_image
+echo "Launching CARDS docker image"
```
```suggestion
echo "Launching CARDS Docker Compose environment"
```

IntegralProgrammer commented on selenium/launch_cards.sh:
```diff
@@ -0,0 +1,48 @@
+echo "Generating docker configuration"
+python3 generate_compose_yaml.py --oak_filesystem --dev_docker_image
+echo "Launching CARDS docker image"
+docker-compose build > /dev/null && docker-compose up -d
```
```suggestion
docker-compose build > /dev/null && docker-compose up -d || exit -1
```

Suppose that some other service, perhaps another CARDS instance, is already running on port `127.0.0.1:8080`. In such a case, `docker-compose up -d` would fail and therefore so should this script otherwise it could incorrectly conclude that CARDS had started because _another_ CARDS instance was already available at `127.0.0.1:8080` - and we want to be sure that we're testing _this_ build.

IntegralProgrammer commented on selenium/launch_cards.sh:
```diff
@@ -0,0 +1,48 @@
+  check_cards_running || handle_cards_java_fail
+  curl --fail http://localhost:8080/system/sling/info.sessionInfo.json > /dev/null 2> /dev/null && break
+  # Fail if cards hasn't started after 10 cycles
+  ((c++)) && ((c==10)) && exit -1
```
Change the variable `c` to something more meaningful, like `cards_readiness_checks` and initialize it to `0` before the start of this loop. It would also be best to instead of hard-code the value `10`, to instead set it as a constant at the beginning of this file.

IntegralProgrammer commented on selenium/pom.xml:
```diff
@@ -0,0 +1,121 @@
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+    </dependency>
```
Shouldn't this also be given a `<scope>test</scope>`?

IntegralProgrammer commented on selenium/pom.xml:
```diff
@@ -0,0 +1,121 @@
+    <dependency>
+     <groupId>io.github.bonigarcia</groupId>
+     <artifactId>webdrivermanager</artifactId>
+      <scope>test</scope>
```
Indentation should be fixed.

IntegralProgrammer commented on selenium/src/test/java/io/uhndata/cards/selenium/SeleniumUtils.java:
```diff
@@ -0,0 +1,62 @@
+    public static WebDriver getDriver()
+    {
+        WebDriver driver = null;
+        WebDriverManager.chromedriver().setup();
```
I wonder if it would be better to use the _Browsers in Docker_ support available in the `io.github.bonigarcia.wdm.WebDriverManager` module - https://github.com/bonigarcia/webdrivermanager#browsers-in-docker. Ideally, you would want the _Browser In Docker_ container to be part of the `compose-cluster` environment and it would access the running CARDS instance at `http://cardsinitial:8080` instead of `http://localhost:8080`. It's just a thought for now, maybe something to do for future work.

sdumitriu replied:
I don't like that we hardcode the browser here. I would like to be able to run the tests on any browser, to catch inconsistencies.

IntegralProgrammer commented on selenium/pom.xml:
```diff
@@ -0,0 +1,121 @@
+     <artifactId>webdrivermanager</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.seleniumhq.selenium</groupId>
+      <artifactId>selenium-chrome-driver</artifactId>
+      <scope>test</scope>
+    </dependency>
```
This has already been declared, so declaring it again generates warnings in the Maven build process.

IntegralProgrammer commented on selenium/pom.xml:
```diff
@@ -0,0 +1,121 @@
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+    </dependency>
+    <dependency>
+     <groupId>io.github.bonigarcia</groupId>
+     <artifactId>webdrivermanager</artifactId>
+      <scope>test</scope>
+    </dependency>
```
Should this be locked to a particular version?

acrowthe replied:
It's locked to a specific version in the root pom.xml:
```
      <dependency>
        <groupId>io.github.bonigarcia</groupId>
        <artifactId>webdrivermanager</artifactId>
        <version>5.4.0</version>
      </dependency>
```

acrowthe committed:
CARDS-59: Add support for running frontend tests using Selenium
Address review comments
- Improve compatibility across different build environments
- Improve maven configuration

sdumitriu committed:
[misc] Fixed bash syntax error

sdumitriu committed:
CARDS-59: Add support for running frontend tests using Selenium
Fixed compilation error

acrowthe committed:
CARDS-59: Add support for running frontend tests using Selenium
Slight log adjustment

acrowthe commented:
Fixes look good: Thanks @sdumitriu

acrowthe committed:
CARDS-59: Add support for running frontend tests using Selenium
Add framework for testing in other browsers

acrowthe commented:
Added framework for testing in other browsers with known issues that will be addressed in separate PRs

IntegralProgrammer commented on selenium/src/test/java/io/uhndata/cards/selenium/SeleniumUtils.java:
```diff
@@ -0,0 +1,138 @@
+        return driver;
+    }
+
+    public static void login(WebDriver driver)
+    {
+        login(driver, "admin", "admin");
+    }
+
+    public static void login(WebDriver driver, String username, String password)
+    {
+        driver.get("http://localhost:8080");
+
+        driver.findElement(By.name("j_username")).sendKeys(username);
+        driver.findElement(By.name("j_password")).sendKeys(password);
+        driver.findElement(By.name("j_password")).submit();
+
+        // TODO: Firefox is not waiting for the dashboard to load before checking for
+        // the avatar so is failing every time. This can be resolved by a sleep() but should be
+        // fixed properly
+        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(5));
+        wait.until(ExpectedConditions.presenceOfElementLocated(By.id("adminnavbaravatar")));
+    }
```
I think it would be better if this code were to be moved into a new file - let's call it `SeleniumCardsUtils.java`. That way, we keep the CARDS specific code in its own file and when we split CARDS into multiple projects, we can create, for example, a `SeleniumHeraclesUtils.java` file which would have methods that use Selenium for testing features that are exclusive to the _HERACLES_ project.

IntegralProgrammer commented on selenium/src/test/java/io/uhndata/cards/selenium/LoginSelenium.java:
```diff
@@ -0,0 +1,45 @@
+public class LoginSelenium
+{
+    private WebDriver driver;
+
```
Would it be a better idea to call `selenium/launch_cards.sh` from here in a `@BeforeClass` method and rename this file to `BasicUiTests.java` where various checks would be performed such as checking the _admin navbar avatar_ but also other checks such as verifying the contents of the sidebar menu items?

acrowthe replied:
Partially implemented: Having to restart cards for every test file will be too slow.
Did rename the file as suggested.

IntegralProgrammer commented on selenium/src/test/java/io/uhndata/cards/selenium/LoginSelenium.java:
```diff
@@ -0,0 +1,45 @@
+    @Test
+    public void loginTest()
+    {
+        this.driver = SeleniumUtils.getDriver();
+        SeleniumUtils.login(this.driver);
```
If we choose to implement the above comment, then this should be moved into the `@BeforeClass` method.

IntegralProgrammer commented on selenium/src/test/java/io/uhndata/cards/selenium/LoginSelenium.java:
```diff
@@ -0,0 +1,45 @@
+        Assert.assertEquals("A", this.driver.findElement(By.id("adminnavbaravatar")).getText());
+    }
+
+    @After
```
If we choose to implement the above two comments, then this should be changed to an `@AfterClass` method which closes the browser and calls `terminate_cards.sh` only after _all_ `@Test` methods have been executed.

IntegralProgrammer reviewed (changes requested):
Code looks pretty good although I have a few small suggestions/ideas that I have discussed in this review.

sashaandjic commented:
![image](https://github.com/data-team-uhn/cards/assets/23243244/8d2a9f3c-6e3e-4eb2-9f28-9d7aff4d5713)

![image](https://github.com/data-team-uhn/cards/assets/23243244/e52ed2c1-8ed3-416b-a4f8-414895b31b96)

acrowthe committed:
CARDS-59: Add support for running frontend tests using Selenium
- Move login to custom TestRule for easy re-use in different tests
- Rename SeleniumUtils to improve naming scheme for future project specific selenium
  util files

acrowthe committed:
CARDS-59: Add support for running frontend tests using Selenium
- Fix Basic Selenium Test name

sashaandjic commented:
ready for re-test?

acrowthe commented:
Yes, it's ready for retest
